### ISSUE TITLE
[Discussion] Creating an Alternative to Goodreads (User Follows, User Review Improvements)

### ISSUE THREAD

**Issue by trosel** [internetarchive/openlibrary#1964]
These are some items that I believe are required to truly make an alternative to Goodreads. 

Perhaps this is beyond the scope of what OpenLibrary is trying to be, but I don't see any point in creating a halfway replacement for Goodreads when the site is already very close. For example, you have the reading lists/logs and star ratings already.

# User Follows and Personalized Feeds

When I create a new account on OpenLibrary, my homepage should have a section for a Feed. 

On an individual's profile today, it shows a feed of *Recent Activity* on the bottom of the page.

When you click a button to "Follow"  a user, your homepage Feed now includes that user's activity.

# User Review Improvements

Today, there are only star reviews (1-5). And I don't believe it's possible to actually see the breakdown of how many of each star there is.

1. Update the data description of a review to be a Star Rating that may or may not include a long text review as well.

2. Create a dedicated page or section for a book's reviews. This would display the star ratings and reviews with links to the review author profiles.

**Comment by mekarpeles** [comment]
Relates to #857, see WIP PR #1966 -- "User Follows" user activity following is a branch-in-progress.

**Comment by trosel** [comment]
Awesome! So for the Rating expansion, this is what I'm thinking:

1. Expand all methods in the [Ratings class](https://github.com/internetarchive/openlibrary/blob/d883f91229b4f4d13fa2e6d4da6343dc78417d64/openlibrary/core/ratings.py) to include: `star_rating` and (optional) `text_review`

2. Search all files for invocations of `rating` and replace it with the appropriate `rating.star_rating`

3. Add a `/book/[id]/ratings` page that will display the breakdown of all of the ratings. [This StarRating.html page](https://github.com/internetarchive/openlibrary/blob/75452cf6f0e3c23e1030395c93e232686ebaa417/openlibrary/macros/StarRatings.html) may be where we want to expand functionality.

4. Begin to add in invocations of `rating.text_review` where wanted/needed

**Comment by LeadSongDog** [comment]
We should distinguish rating the work from rating the various editions, especially for translations and audiobooks.  Not all readings are equivalent.

**Comment by trosel** [comment]
@LeadSongDog If we could have the option for edition vs work rating (or both), that might be interesting. 

I often see amazon book reviews that logically separate the edition from the general work.

**Comment by seabelis** [comment]
This has the potential of creating "edit wars" - especially for current books. A book's reviews can be boosted by changing the details of an edition. Might not be a deal-breaker for the feature, but may create the need for some written policies and moderation.

**Comment by trosel** [comment]
@seabelis can you give an example of what you're talking about? Please use the words "work" and "edition" so we can understand your concern.

**Comment by tfmorris** [comment]
Goodreads is a commercial social site for readers. OpenLibrary is a non-profit site with the mission statement "A web page for every book." Those are two very different things and OpenLibrary is a long way from doing a good job at its core mission, so I think it's way too early to try and warp it into something else.

If we were to compete with Goodreads, the first thing we should do is stop sending them traffic, as we currently do.

Text reviews require spam filtering, moderation, flagging of inappropriate reviews, and a whole bunch of related infrastructure.

A followers feature requires privacy controls (which #1966 blithely says will be "later"), dealing with follower spam, etc.

Both of these require ramping of manpower, whether it be paid or volunteer.

All of OpenLibrary's data is freely available, so if someone wants to build a Goodreads competitor, they have all the data that they need to do it.

**Comment by trosel** [comment]
>Both of these require ramping of manpower, whether it be paid or volunteer.

This is irrelevant, imo. All it takes is one dedicated person to come along. I'm not demanding that other people do something immediately. 

>All of OpenLibrary's data is freely available, so if someone wants to build a Goodreads competitor, they have all the data that they need to do it.

Good point!

**Comment by theashwin** [comment]
Can this be taken up as a Google Summer of Code proposal? If yes, then it'll be an interesting project for the summer. I have some ideas on how I would implement these features. @trosel

**Comment by mekarpeles** [comment]
@theashwin a good place to start with a Google summer of code proposal is learning from the community what their thoughts are values are. @tfmorris has been a core contributor for several years and I agree with a lot of his points here. I'm not yet convinced these features are the right ones for open Library and I'm not sure how the rest of our community would respond to a Google summer of code proposal on this topic. You're welcome to bring it to our community call for feedback (there's one today at 11:30am PT)

**Comment by mekarpeles** [comment]
@trosel /book/[id]/ratings seems like a reasonable proposal to me.

**Comment by jdlrobson** [comment]
> This is irrelevant, imo. All it takes is one dedicated person to come along. I'm not demanding that other people do something immediately.

So what happens when that dedicated person disappears or has less time to focus on the feature(as is common in open source)? Who reviews their code while they are here? Who is on the hook when the site goes down because of code relating to those changes?

I'm inclined to agree with @tfmorris on this one

**Comment by jdlrobson** [comment]
My biggest concern around text reviews is moderation is a huge problem and spam, copyright, harassment and offensive text are all things you should be thinking about with such a system. 

Who will moderate? Can the community handle that? I'd like to see these in the proposal.

I think the goal is admirable but we should consider every possible thing that is packed into that goal.

**Comment by seabelis** [comment]
> @seabelis can you give an example of what you're talking about? Please use the words "work" and "edition" so we can understand your concern.

The terms make no difference as both items can be freely edited by anyone. The danger in adding reviews is that they can be easily hijacked by editing the details of a work or edition record. Or on the flip side to disassociate a work or edition from poor reviews. And this also provides incentive to modify records for reasons other than improving the quality of the data itself.

**Comment by trosel** [comment]
@jdlrobson 

>So what happens when that dedicated person disappears or has less time to focus on the feature(as is common in open source)? Who reviews their code while they are here? Who is on the hook when the site goes down because of code relating to those changes?

This is a concern for literally any work that goes into the site. Again, there's no need to rush into anything. But there's also no need to dismiss ideas merely because an Open Source project is Open Source.

@mekarpeles @tfmorris My question based on your concerns is where the line gets drawn. Adding the User Follows / Activity feeds is about 70% of the way towards Goodreads. Is _that_ where we want to draw the line in terms of similarity? I just came across this project, so I'm open to understanding more about what OpenLibrary is designed to be and what it is not designed to be.

**Comment by brad2014** [comment]
@mek do you want to wrangle the community toward a decision on how to proceed with this?  I'm not clear on how this issue gets closed (or morphed into specific tracking issues).

**Comment by guyjeangilles** [comment]
@mekarpeles doubling down on what brad said about making a decision. I think the discussion her is lukewarm at best with regards to making a alternative to Goodreads. I think there are enough people opposed that this issue should be closed. I for one think Open Library should focus on having abundant data on books, not competing with any particular site.

**Comment by trosel** [comment]
@guyjeangilles Is OpenLibrary primarily meant to be an API?

If so, can you address my question here:

>Adding the User Follows / Activity feeds is about 70% of the way towards Goodreads. Is that where we want to draw the line in terms of similarity?

**Comment by guyjeangilles** [comment]
tl;dr this kinda turned into a novel but basically Open Library is a library and Goodreads isn't.

@trosel I'm not the one to definitively say what Open Library is or isn't, but lets look at what each site says about itself to consider where "the line of similarity" is.

[From Goodreads about us](https://www.goodreads.com/about/us):
>Goodreads is the world’s largest site for readers and book recommendations. Our mission is to help people find and share books they love. Goodreads launched in January 2007.
**A Few Things You Can Do On Goodreads**
>
 >   See what books your friends are reading.
  >  Track the books you're reading, have read, and want to read.
    Check out your personalized book recommendations. Our recommendation engine analyzes 20 billion data points to give suggestions tailored to your literary tastes.
   Find out if a book is a good fit for you from our community’s reviews.

I would describe Goodreads as a social book catalog. I'd also very much describe Goodreads as for-profit. One of Goodreads goals is to drive traffic to Amazon and generate revenue from Goodreads Author profiles. As tfmorris mentioned above, if we were to truly make a competitor to Goodreads, we should stop driving traffic to them, and I believe Open Library should not restrict linking to other quality websites.

Now consider Open Library's Vision.

[From Open Library's Vision page](https://openlibrary.org/about/vision):
>The ultimate goal of the Open Library is to make all the published works of humankind available to everyone in the world. While large in scope and ambition, this goal is within our grasp. Achieving it will require the participation of librarians, authors, government officials and technologists.
>...
>How can you help? Improve the records we have for the books you love – every record is fully editable by clicking the “edit” button on the page. If we don't have a record for a book, you can create one. If you're a library with digitized resources, tell us where they are and we'll point people to them.
>
>Our hope is that libraries and individual book readers will join this project and together we can build towards universal access to all knowledge.

So in a long winded way to answer your question: Open Library, from it's own vision, is not an API. It's a wiki. It's quite clear that the goal of Open Library is knowledge and not necessarily the social aspects of reading. It's also very clear that Open Library is indeed a library and one that is accredited by the State of California. Libraries and social book catalogs (which Goodreads is) have distinct functions. Particularly, libraries are about archiving knowledge and making that knowledge readily available to it's patrons.

Instead of asking "what is the line between Goodreads and Open Library", for every feature proposed we should ask: 1) "How does this increase the knowledge Open Library contains" and/or 2) "How does this feature/enhancement improve the ability of users to access said knowledge."

Sometimes there will be features that fit Open Library's vision and overlap with Goodreads' functionality. This isn't necessarily evidence that Open Library can become an alternative to Goodreads. The goals of the two sites are distinct and should not be conflated, in my opinion.

**Comment by trosel** [comment]
@guyjeangilles That is helpful. But unfortunately, there are some features in Open Library that betray that vision.

For example, why is there a reading log with `Currently Reading/Want to Read/Already Read` lists?

**Comment by guyjeangilles** [comment]
@trosel Again, I'm not the one who defines Open Library's vision, but you could argue that a principle function of a library is to facilitate book lending for it's patrons. Many popular digital library apps (i.e Overdrive, Libby) provide features that allow their users to make `Currently Reading/Want to Read/Already Read` as well as `Liked`, `Disliked`, etc. 

Someone who wanted to make a Goodreads alternative could make arguments for including the same features in their site. This overlap is not important. Want's important is the _intent_ behind the feature. If the intent is to be an alternative to Goodreads...well Open Library's data is open source and anyone could make an alternative if they so choose. I believe Open Library should stay a library focused on archiving knowledge and making it readily available to everyone and anyone.

**Comment by trosel** [comment]
@guyjeangilles Thanks for the feedback. I think we're on the same page because what this issue is suggesting is exactly what you explained here:

> Many popular digital library apps (i.e Overdrive, Libby) provide features that allow their users to make Currently Reading/Want to Read/Already Read as well as Liked, Disliked, etc.

**Comment by guyjeangilles** [comment]
@trosel 

> ...I think we're on the same page because what this issue is suggesting is exactly what you explained here:
> 
> > Many popular digital library apps (i.e Overdrive, Libby) provide features that allow their users to make Currently Reading/Want to Read/Already Read as well as Liked, Disliked, etc.

If this is true than the original issue is poorly worded and poorly titled. (Quote emphasis mine)
> These are some items that I believe are required to truly make _**an alternative to Goodreads.**_

I personally don't mind Open Library eventually becoming an alternative to Overdrive or Libby. If that's the consensus we've reached, then I'd change the title of the issue and leave it open.

**Comment by trosel** [comment]
@guyjeangilles When I think of Goodreads, I don't think "this is primarily intended to drive revenue for amazon". I think "this is a book wiki with a social network laid on top of it". 

Goodreads is kind of a household name for better or worse and that's why I used it. In any case, I think this is a nightmare and I'm happy to close.

**Comment by LongJohn-Silver** [comment]
> Goodreads is a commercial social site for readers. OpenLibrary is a non-profit site with the mission statement "A web page for every book." Those are two very different things and OpenLibrary is a long way from doing a good job at its core mission, so I think it's way too early to try and warp it into something else.
> 
> If we were to compete with Goodreads, the first thing we should do is stop sending them traffic, as we currently do.
> 
> Text reviews require spam filtering, moderation, flagging of inappropriate reviews, and a whole bunch of related infrastructure.
> 
> A followers feature requires privacy controls (which #1966 blithely says will be "later"), dealing with follower spam, etc.
> 
> Both of these require ramping of manpower, whether it be paid or volunteer.
> 
> All of OpenLibrary's data is freely available, so if someone wants to build a Goodreads competitor, they have all the data that they need to do it.

I messaged a lot of my contacts in GoodReads to see if they would switch to OL and most of them said that it lacks social features, I think that if that kind of things would be added more users would come, and sure, more spam and bots maybe, but also people willing to help with spam control and a better cataloguing and else. Regarding being a non-profit, I don't think that changes anything to become an alternative, even if it's not as good now.



> My biggest concern around text reviews is moderation is a huge problem and spam, copyright, harassment and offensive text are all things you should be thinking about with such a system.
> 
> Who will moderate? Can the community handle that? I'd like to see these in the proposal.
> 
> I think the goal is admirable but we should consider every possible thing that is packed into that goal.

I agree in terms of spam, can't say much about the code maintaining, but I guess a good starting point would be librarians who volunteer to handle spam, capchas for every review (you could use something like Lichess captcha maybe?) and limiting the amount of reviews to a certain time, and else, I'm not saying it's super easy, but I think it's doable and I volunteer myself to help with spam if that's something.
